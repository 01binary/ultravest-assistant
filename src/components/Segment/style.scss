@import '../App/style/colors';
@import '../App/style/layout';

$segment-width: 160px;
$segment-height: 62px;
$segment-overlap: -34px;
$segment-icon-size: 35px;
$segment-extend: $segment-icon-size + $spacing-three-halves;

// Segment wrapper
.segment {
	padding: $spacing-third;
}

// Segment content
.segment article {
	position: relative;
	background-repeat: no-repeat;
	width: $segment-width;
	height: $segment-height;
	margin-right: $segment-overlap;
}

// Last segment content
.segment:last-of-type article {
	margin-right: 0;
}

// Segment raising diagram
.raise {
	background-image: url('../../assets/raise.svg');
}

// Segment lowering diagram
.lower {
	background-image: url('../../assets/lower.svg');
}

// Segment tip icon
.segment figure {
	position: absolute;
	cursor: pointer;
	z-index: 2;
	background-repeat: no-repeat;
}

// Segment ambient temperature overlay
.ambient {
	width: 28px;
	height: $spacing-single;
	top: 24px;
	left: -43px;
}

// Segment temperature icon
.temperature {
	background-image: url('../../assets/pyrometer.svg');

	// Icon width
	width: $segment-icon-size;

	// Extend to cover graph point
	height: 60px;
}

// Segment time icon
.time {
	background-image: url('../../assets/clock.svg');

	// Icon width
	width: $segment-icon-size;
	
	// Extend to cover graph point
	height: 60px;
}

// Segment rate icon
.rate {
	background-image: url('../../assets/up.svg');
	width: $segment-extend;
	height: $segment-icon-size;
	left: 0px;
	top: -10px;
}

// Segment temperature up icon
.raise .temperature {
	left: 47px;
	top: -55px;
}

// Segment time up icon
.raise .time {
	left: 93px;
	top: -55px;
}

// Segment rate down icon
.lower .rate {
	background-image: url('../../assets/down.svg');
	left: -2px;
	top: $spacing-third;
}

// Segment temperature down icon
.lower .temperature {
	left: 47px;
	top: $spacing-single;
	background-position-y: $spacing-single + $spacing-min;
	height: $segment-icon-size + $spacing-single + $spacing-min;
}

// Segment time down icon
.lower .time {
	left: 93px;
	top: $spacing-single;
	background-position-y: $spacing-single + $spacing-min;
	height: $segment-icon-size + $spacing-single + $spacing-min;
}

// Segment tip
.segment figcaption {
	position: absolute;
	line-height: $spacing-single;
	background: $color-border-light;
	padding: $spacing-third $spacing-half;
	min-width: 4em;
	bottom: calc(100% + 1em);
	right: 1em;
	opacity: 0;
	transform: translateY($spacing-third);
	transition:
		opacity .2s ease-in-out,
		transform .2s ease-in-out;
	pointer-events: none;
}

// Offset segment tip for rate
.rate figcaption {
	right: 2em;
}

// Segment tip hover
.segment figure:hover > figcaption {
	opacity: 1;
	transform: translateY(0);
}

// Segment tip point
.segment figcaption:before {
	content: '';
	position: absolute;
	top: 100%;
	right: 0px;
	border-width: $spacing-quarter;
	border-style: solid;
	border-color:
		$color-border-light
		$color-border-light
		transparent
		transparent;
}

// Segment tip down
.segment .lower figcaption {
	top: calc(100% + 1em);
	bottom: initial;
}

// Segment down tip arrow
.segment .lower figcaption:before {
	top: -$spacing-half;
	border-color:
		transparent
		$color-border-light
		$color-border-light
		transparent;
}

// Segment tip heading
.segment h3 {
	font-size: $spacing-four-thirds;
	margin: 0;
	line-height: $spacing-single;
}

// Units span
.units {
	color: $color-text-light;
}
