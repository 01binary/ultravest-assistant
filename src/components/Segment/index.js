import { h } from 'preact';
import classNames from 'obj-str';
import changeCase from 'change-case';
import style from './style';

/**
 * Burnout diagram segment.
 * @param {string} name - The segment name.
 * @param {number} rate - The segment rate (F per hour).
 * @param {number} temp - The segment temperature (F).
 * @param {number} hold - The segment hold time (hours).
 * @param {Object} prev - The previous segment.
 */
const Segment = ({
	name,
	rate,
	temp,
	hold,
	prev
}) => (
	<a name={getSegmentAnchor(name)}>
		<article class={getSegmentClass(prev && prev.temp, temp)}>
			<figure class={style.temperature}>
				<svg />

				<figcaption>
					<h3>temp</h3>
					{ prev ?
						`${temp}°` : 'ambient'
					}
					{ prev &&
						<span class={style.units}>F</span>
					}
				</figcaption>
			</figure>

			<figure class={style.rate}>
				<svg />

				<figcaption>
					<h3>rate</h3>
					{ `${rate}°` }
					<span class={style.units}>F/hr</span>
				</figcaption>
			</figure>

			<figure class={style.time}>
				<svg />

				<figcaption>
					<h3>hold</h3>
					{hold}
					<span class={style.units}>
						{hold > 1 ? 'hours' : 'hour'}
					</span>
				</figcaption>
			</figure>
		</article>
	</a>
);

const getSegmentClass = (prevTemp, temp) => classNames({
	[style.segment]: true,
	[style.raise]: prevTemp ? temp > prevTemp : true,
	[style.lower]: prevTemp ? temp < prevTemp : false
});

const getSegmentAnchor = name => (
	changeCase.paramCase(name)
);

export default Segment;
