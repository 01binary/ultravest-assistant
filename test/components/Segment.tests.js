/*import { h } from 'preact';
import { deep } from 'preact-render-spy';
import Segment from '../../src/components/Segment';
import presets from '../../src/config/burnoutPresets';

describe('Segment', () => {

	let wrapper;

	beforeAll(() => {
		wrapper = deep(
			<Segment {...props} />
		);
	});

	afterAll(() => {
		wrapper = null;
	});

	test('should render', () => {
		expect(wrapper).toMatchSnapshot();
	});

	const first = Object.keys(presets)[0];
	const segments = presets[first].segments;
	const cur = segments[1];
	const prev = segments[0];
	const props = Object.assign({}, cur, { prev });
});
*/