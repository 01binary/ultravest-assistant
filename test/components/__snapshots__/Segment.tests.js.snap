// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Segment should render 1`] = `
preact-render-spy (1 nodes)
-------
<a name="thermal-transition">
  <article class="segment raise">
    <figure class="temperature">
      <svg></svg>
      <figcaption>
        <h3>temp</h3>
        700°
        <span class="units">F</span>
      </figcaption>
    </figure>
    <figure class="rate">
      <svg></svg>
      <figcaption>
        <h3>rate</h3>
        400°
        <span class="units">F/hr</span>
      </figcaption>
    </figure>
    <figure class="time">
      <svg></svg>
      <figcaption>
        <h3>hold</h3>
        1
        <span class="units">hour</span>
      </figcaption>
    </figure>
  </article>
</a>

`;
